tests/integration/src/disaster_recovery_tests.rs

```diff
@@ -997,3 +1009,182 @@ fn test_disaster_recovery_failing() {
+    match dr_status.last_recovery_result {
+        Some(upgrader_api::RecoveryResult::Failure(err)) => {
+            assert!(err.reason.contains("Stop canister request timed out"))
+        }
+        _ => panic!(
+            "Unexpected last recovery result: {:?}",
+            dr_status.last_recovery_result
+        ),
+    };
+
+    // the station should still be "Stopping"
+    let station_status = env
+        .canister_status(canister_ids.station, Some(upgrader_id))
+        .unwrap();
+    assert!(matches!(
+        station_status.status,
+        CanisterStatusResultStatus::Stopping
+    ));
+
+    // force stop in disaster recovery
+    disaster_recovery_request.force = true;
+
+    // we perform successful disaster recovery with forced stop twice
+    // to test that snapshots can be taken multiple times
+    let mut snapshots = env
+        .list_canister_snapshots(canister_ids.station, Some(upgrader_id))
+        .unwrap();
+    assert!(snapshots.is_empty());
+    for i in 0..2 {
+        let new_name = format!("recovered-{}", i);
+        let station_init_arg = SystemInstall::Upgrade(SystemUpgrade {
+            name: Some(new_name.clone()),
+        });
+        disaster_recovery_request.arg = Encode!(&station_init_arg).unwrap();
+        let res: (ApiResult<()>,) = update_candid_as(
+            &env,
+            upgrader_id,
+            WALLET_ADMIN_USER,
+            "request_disaster_recovery",
+            (disaster_recovery_request.clone(),),
+        )
+        .unwrap();
+        res.0.unwrap();
+        // start processing the mgmt canister call from upgrader to stop the station
+        env.tick();
+
+        // the station can't be stopped yet because it has an open call context
+        // with a pending down-stream call to the "expensive" method of the test canister
+        // now we advance time by 5 mins to time out (i.e., fail) the upgrader's call to stop the station
+        env.advance_time(Duration::from_secs(5 * 60));
+
+        // disaster recovery should succeed now when forcing the station to stop
+        await_disaster_recovery_success(&env, canister_ids.station, upgrader_id);
+
+        // check that a snapshot has been taken
+        let current_snapshots = env
+            .list_canister_snapshots(canister_ids.station, Some(upgrader_id))
+            .unwrap();
+        assert_eq!(current_snapshots.len(), 1);
+        assert!(
+            snapshots.is_empty()
+                || (snapshots.len() == 1 && current_snapshots[0].id != snapshots[0].id)
+        );
+        snapshots = current_snapshots;
+
+        // check new station name set during disaster recovery
+        let system_info = get_system_info(&env, WALLET_ADMIN_USER, canister_ids.station);
+        assert_eq!(system_info.name, new_name);
+
+        // the call request will be "Processing" forever since we deleted its call context during disaster recovery
```

Comment:
not related to this PR, but we should ideally have something in place to perform some cleanup on requests that are forever `processing` since they are actually not processing, it's just dangling state